File: Services/SubscriptionService.swift
import Foundation
import StoreKit
import Combine
import FirebaseFirestore
import FirebaseAuth
import FirebaseAnalytics

class SubscriptionService: NSObject, ObservableObject, SKPaymentTransactionObserver {
    static let shared = SubscriptionService()
    
    @Published var products: [SKProduct] = []
    @Published var isSubscribed: Bool = false
    @Published var lastPurchasedProduct: SKProduct?
    @Published var isLoading: Bool = false
    @Published var errorMessage: String?
    
    private let productIdentifiers: Set<String> = [
        "com.lyes.love2love.subscription.weekly",
        "com.lyes.love2love.subscription.monthly"
    ]
    
    private var cancellables = Set<AnyCancellable>()
    
    override init() {
        super.init()
        SKPaymentQueue.default().add(self)
        
        // Nettoyer les transactions en attente au démarrage pour éviter les blocages
        clearPendingTransactions()
        
        // Charger les produits après un court délai pour éviter les conflits
        DispatchQueue.main.asyncAfter(deadline: .now() + 1.0) {
            self.loadProducts()
        }
        
        checkSubscriptionStatus()
    }
    
    deinit {
        SKPaymentQueue.default().remove(self)
    }
    
    func loadProducts() {
        print("🔥 SubscriptionService: Chargement des produits...")
        
        let request = SKProductsRequest(productIdentifiers: productIdentifiers)
        request.delegate = self
        request.start()
        
        // Timeout de sécurité
        DispatchQueue.main.asyncAfter(deadline: .now() + 30) { [weak self] in
            if self?.products.isEmpty == true {
                print("⚠️ SubscriptionService: Timeout - Aucune réponse d'Apple après 30s")
                DispatchQueue.main.async {
                    self?.errorMessage = "Impossible de charger les offres d'abonnement. Vérifiez votre connexion et réessayez."
                }
            }
        }
    }
    
    func purchase(product: SKProduct) {
        print("🔥 SubscriptionService: Tentative d'achat: \(product.productIdentifier)")
        
        guard SKPaymentQueue.canMakePayments() else {
            print("❌ SubscriptionService: Achats non autorisés sur cet appareil")
            errorMessage = "Les achats ne sont pas autorisés sur cet appareil"
            return
        }
        
        isLoading = true
        let payment = SKPayment(product: product)
        SKPaymentQueue.default().add(payment)
    }
    
    func restorePurchases() {
        print("🔥 SubscriptionService: Restauration des achats...")
        
        isLoading = true
        SKPaymentQueue.default().restoreCompletedTransactions()
    }
    
    /// Nettoyer les transactions en attente qui peuvent bloquer les achats
    private func clearPendingTransactions() {
        let pendingTransactions = SKPaymentQueue.default().transactions
        
        guard !pendingTransactions.isEmpty else { return }
        
        print("🔧 SubscriptionService: Nettoyage de \(pendingTransactions.count) transaction(s) en attente")
        
        for transaction in pendingTransactions {
            switch transaction.transactionState {
            case .purchased, .restored, .failed:
                SKPaymentQueue.default().finishTransaction(transaction)
            case .purchasing:
                print("⚠️ SubscriptionService: Transaction en cours d'achat détectée")
            case .deferred:
                print("⚠️ SubscriptionService: Transaction différée détectée")
            @unknown default:
                break
            }
        }
        
        // Réinitialiser l'état de chargement
        DispatchQueue.main.async {
            self.isLoading = false
            self.errorMessage = nil
        }
    }
    
    private func checkSubscriptionStatus() {
        // Vérifier le statut d'abonnement depuis Firebase
        if let user = FirebaseService.shared.currentUser {
            isSubscribed = user.isSubscribed
        }
    }
    
    // MARK: - SKPaymentTransactionObserver
    
    func paymentQueue(_ queue: SKPaymentQueue, updatedTransactions transactions: [SKPaymentTransaction]) {
        print("🔥 SubscriptionService: Mise à jour des transactions - \(transactions.count) transaction(s)")
        NSLog("🔥 SubscriptionService: Mise à jour des transactions - \(transactions.count) transaction(s)")
        
        for transaction in transactions {
            print("🔥 SubscriptionService: Transaction \(transaction.transactionIdentifier ?? "unknown") - État: \(transaction.transactionState.rawValue)")
            NSLog("🔥 SubscriptionService: Transaction \(transaction.transactionIdentifier ?? "unknown") - État: \(transaction.transactionState.rawValue)")
            
            switch transaction.transactionState {
            case .purchased:
                print("🔥 SubscriptionService: Transaction PURCHASED")
                NSLog("🔥 SubscriptionService: Transaction PURCHASED")
                handleSuccessfulPurchase(transaction)
            case .restored:
                print("🔥 SubscriptionService: Transaction RESTORED")
                NSLog("🔥 SubscriptionService: Transaction RESTORED")
                handleRestored(transaction)
            case .failed:
                print("🔥 SubscriptionService: Transaction FAILED")
                NSLog("🔥 SubscriptionService: Transaction FAILED")
                handleFailed(transaction)
            case .deferred:
                print("🔥 SubscriptionService: Transaction DEFERRED")
                NSLog("🔥 SubscriptionService: Transaction DEFERRED")
            case .purchasing:
                print("🔥 SubscriptionService: Transaction PURCHASING - Sheet Apple devrait apparaître")
                NSLog("🔥 SubscriptionService: Transaction PURCHASING - Sheet Apple devrait apparaître")
                // S'assurer que le loading reste actif pendant l'achat
                DispatchQueue.main.async {
                    self.isLoading = true
                }
                
                // Timeout de sécurité pour éviter le chargement infini
                DispatchQueue.main.asyncAfter(deadline: .now() + 60) { [weak self] in
                    // Vérifier si la transaction est toujours en cours après 60s
                    if let currentTransaction = queue.transactions.first(where: { $0.transactionIdentifier == transaction.transactionIdentifier }),
                       currentTransaction.transactionState == .purchasing {
                        print("⚠️ SubscriptionService: Transaction bloquée en état purchasing depuis 60s")
                        NSLog("⚠️ SubscriptionService: Transaction bloquée en état purchasing depuis 60s")
                        DispatchQueue.main.async {
                            self?.isLoading = false
                            self?.errorMessage = "La transaction a pris trop de temps. Veuillez réessayer."
                        }
                    }
                }
            @unknown default:
                print("🔥 SubscriptionService: Transaction état inconnu: \(transaction.transactionState.rawValue)")
                NSLog("🔥 SubscriptionService: Transaction état inconnu: \(transaction.transactionState.rawValue)")
                break
            }
        }
    }
    
    private func handleSuccessfulPurchase(_ transaction: SKPaymentTransaction) {
        print("🔥 SubscriptionService: ✅ Achat réussi: \(transaction.payment.productIdentifier)")
        NSLog("🔥 SubscriptionService: ✅ Achat réussi: \(transaction.payment.productIdentifier)")
        
        // 📊 Analytics: Abonnement réussi avec le bon type
        let planType = transaction.payment.productIdentifier.contains("weekly") ? "weekly" : "monthly"
        Analytics.logEvent("abonnement_reussi", parameters: [
            "type": planType,
            "source": "storekit_success"
        ])
        print("📊 Événement Firebase: abonnement_reussi - type: \(planType) - source: storekit_success")
        
        isSubscribed = true
        isLoading = false
        
        // NOUVEAU: Marquer comme abonnement direct AVANT de mettre à jour Firebase
        Task {
            await markSubscriptionAsDirect()
            
            // Mettre à jour Firebase APRÈS avoir marqué l'abonnement comme direct
            await MainActor.run {
                FirebaseService.shared.updateSubscriptionStatus(isSubscribed: true)
            }
            
            // Le partage automatique sera géré par PartnerSubscriptionSyncService
            print("🔥 SubscriptionService: Synchronisation automatique via PartnerSubscriptionSyncService")
        }
        
        SKPaymentQueue.default().finishTransaction(transaction)
        
        // Notifier le succès
        NotificationCenter.default.post(
            name: NSNotification.Name("SubscriptionPurchased"),
            object: nil
        )
        
        print("🔥 SubscriptionService: Achat finalisé et partagé avec partenaire")
        NSLog("🔥 SubscriptionService: Achat finalisé et partagé avec partenaire")
    }
    
    private func handleRestored(_ transaction: SKPaymentTransaction) {
        print("🔥 SubscriptionService: ✅ Achat restauré: \(transaction.payment.productIdentifier)")
        NSLog("🔥 SubscriptionService: ✅ Achat restauré: \(transaction.payment.productIdentifier)")
        
        // 📊 Analytics: Achat restauré
        Analytics.logEvent("achat_restaure", parameters: [:])
        print("📊 Événement Firebase: achat_restaure")
        
        isSubscribed = true
        isLoading = false
        
        // NOUVEAU: Marquer comme abonnement direct AVANT de mettre à jour Firebase
        Task {
            await markSubscriptionAsDirect()
            
            // Mettre à jour Firebase APRÈS avoir marqué l'abonnement comme direct
            await MainActor.run {
                FirebaseService.shared.updateSubscriptionStatus(isSubscribed: true)
            }
            
            // Le partage automatique sera géré par PartnerSubscriptionSyncService
            print("🔥 SubscriptionService: Synchronisation automatique via PartnerSubscriptionSyncService")
        }
        
        SKPaymentQueue.default().finishTransaction(transaction)
        
        // Notifier la restauration réussie
        NotificationCenter.default.post(
            name: NSNotification.Name("SubscriptionRestored"),
            object: nil
        )
        
        print("🔥 SubscriptionService: Restauration finalisée et partagée avec partenaire")
        NSLog("🔥 SubscriptionService: Restauration finalisée et partagée avec partenaire")
    }
    
    private func handleFailed(_ transaction: SKPaymentTransaction) {
        print("🔥 SubscriptionService: ❌ Transaction échouée")
        NSLog("🔥 SubscriptionService: ❌ Transaction échouée")
        
        isLoading = false
        
        if let error = transaction.error as? SKError {
            print("🔥 SubscriptionService: Code d'erreur SKError: \(error.code.rawValue)")
            NSLog("🔥 SubscriptionService: Code d'erreur SKError: \(error.code.rawValue)")
            
            if error.code != .paymentCancelled {
                print("🔥 SubscriptionService: ❌ Erreur d'achat: \(error.localizedDescription)")
                NSLog("🔥 SubscriptionService: ❌ Erreur d'achat: \(error.localizedDescription)")
                errorMessage = "Erreur d'achat: \(error.localizedDescription)"
            } else {
                print("🔥 SubscriptionService: Achat annulé par l'utilisateur")
                NSLog("🔥 SubscriptionService: Achat annulé par l'utilisateur")
            }
        } else if let error = transaction.error {
            print("🔥 SubscriptionService: ❌ Erreur générale: \(error.localizedDescription)")
            NSLog("🔥 SubscriptionService: ❌ Erreur générale: \(error.localizedDescription)")
        }
        
        SKPaymentQueue.default().finishTransaction(transaction)
        print("🔥 SubscriptionService: Transaction échouée finalisée")
        NSLog("🔥 SubscriptionService: Transaction échouée finalisée")
    }
    
    func paymentQueue(_ queue: SKPaymentQueue, restoreCompletedTransactionsFailedWithError error: Error) {
        print("🔥 SubscriptionService: ❌ Erreur de restauration: \(error.localizedDescription)")
        NSLog("🔥 SubscriptionService: ❌ Erreur de restauration: \(error.localizedDescription)")
        
        isLoading = false
        errorMessage = "Erreur de restauration: \(error.localizedDescription)"
    }
    
    func paymentQueueRestoreCompletedTransactionsFinished(_ queue: SKPaymentQueue) {
        print("🔥 SubscriptionService: ✅ Restauration terminée - \(queue.transactions.count) transaction(s)")
        NSLog("🔥 SubscriptionService: ✅ Restauration terminée - \(queue.transactions.count) transaction(s)")
        
        isLoading = false
        if queue.transactions.isEmpty {
            errorMessage = "Aucun achat à restaurer trouvé"
            print("🔥 SubscriptionService: Aucun achat à restaurer")
            NSLog("🔥 SubscriptionService: Aucun achat à restaurer")
        }
    }
    
    // MARK: - Gestion de la résiliation d'abonnement
    
    func handleSubscriptionExpired() async {
        guard let currentUser = Auth.auth().currentUser else {
            print("🔥 SubscriptionService: Pas d'utilisateur connecté pour résiliation")
            return
        }
        
        print("🔥 SubscriptionService: Gestion résiliation abonnement pour: \(currentUser.uid)")
        
        // Révoquer l'abonnement de l'utilisateur principal
        await revokeUserSubscription(userId: currentUser.uid)
        
        // Mettre à jour l'état local
        await MainActor.run {
            self.isSubscribed = false
        }
        
        // Mettre à jour Firebase
        FirebaseService.shared.updateSubscriptionStatus(isSubscribed: false)
        
        // La synchronisation avec le partenaire sera automatiquement gérée par PartnerSubscriptionSyncService
        print("✅ SubscriptionService: Résiliation effectuée, synchronisation automatique en cours")
    }
    
    private func revokeUserSubscription(userId: String) async {
        do {
            try await Firestore.firestore()
                .collection("users")
                .document(userId)
                .updateData([
                    "isSubscribed": false,
                    "subscriptionExpiredAt": Timestamp(date: Date()),
                    "subscriptionType": FieldValue.delete()
                ])
            
            print("✅ SubscriptionService: Abonnement révoqué pour utilisateur: \(userId)")
            
        } catch {
            print("❌ SubscriptionService: Erreur révocation utilisateur: \(error)")
        }
    }
    
    // MARK: - Vérification périodique des abonnements
    
    func startSubscriptionStatusMonitoring() {
        // Timer pour vérifier l'état des abonnements toutes les heures
        Timer.scheduledTimer(withTimeInterval: 3600, repeats: true) { _ in
            Task {
                await self.checkSubscriptionStatus()
            }
        }
        
        print("🔥 SubscriptionService: Monitoring des abonnements démarré")
    }
    
    private func checkSubscriptionStatus() async {
        guard let currentUser = Auth.auth().currentUser else { return }
        
        print("🔥 SubscriptionService: Vérification statut abonnement...")
        
        // TODO: Ici vous devriez implémenter la vérification Apple Receipt
        // Pour l'instant, simulation basée sur l'état Firebase
        
        do {
            let userDoc = try await Firestore.firestore()
                .collection("users")
                .document(currentUser.uid)
                .getDocument()
            
            guard let userData = userDoc.data(),
                  let isSubscribed = userData["isSubscribed"] as? Bool else { return }
            
            // Si l'utilisateur était abonné localement mais ne l'est plus dans Firebase
            if self.isSubscribed && !isSubscribed {
                print("🔥 SubscriptionService: Détection de résiliation d'abonnement")
                await handleSubscriptionExpired()
            }
            
        } catch {
            print("❌ SubscriptionService: Erreur vérification statut: \(error)")
        }
    }
    
    private func markSubscriptionAsDirect() async {
        guard let currentUser = Auth.auth().currentUser else { return }
        
        do {
            try await Firestore.firestore()
                .collection("users")
                .document(currentUser.uid)
                .updateData([
                    "subscriptionType": "direct",
                    "subscriptionPurchasedAt": Timestamp(date: Date())
                ])
            
            print("✅ SubscriptionService: Abonnement marqué comme direct")
            
        } catch {
            print("❌ SubscriptionService: Erreur marquage abonnement direct: \(error)")
        }
    }
}

// MARK: - SKProductsRequestDelegate

extension SubscriptionService: SKProductsRequestDelegate {
    func productsRequest(_ request: SKProductsRequest, didReceive response: SKProductsResponse) {
        print("🔥 SubscriptionService: Réponse reçue - \(response.products.count) produits valides")
        
        if !response.invalidProductIdentifiers.isEmpty {
            print("❌ SubscriptionService: Identifiants invalides: \(response.invalidProductIdentifiers)")
        }
        
        DispatchQueue.main.async {
            self.products = response.products
            
            if self.products.isEmpty {
                print("❌ SubscriptionService: Aucun produit chargé")
                self.errorMessage = "Aucun produit disponible. Vérifiez votre connexion."
            }
        }
    }
    
    func request(_ request: SKRequest, didFailWithError error: Error) {
        print("❌ SubscriptionService: Erreur chargement produits: \(error.localizedDescription)")
        
        DispatchQueue.main.async {
            self.errorMessage = "Erreur lors du chargement des produits: \(error.localizedDescription)"
        }
    }
} 

File: Services/SubscriptionService.kt
// SubscriptionService.kt
package com.lyes.love2love.billing

import android.app.Activity
import android.content.Context
import android.util.Log
import com.android.billingclient.api.*
import com.google.firebase.analytics.ktx.analytics
import com.google.firebase.auth.ktx.auth
import com.google.firebase.firestore.FieldValue
import com.google.firebase.firestore.ktx.firestore
import com.google.firebase.ktx.Firebase
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import java.util.concurrent.atomic.AtomicBoolean

/**
 * Réécriture Android de SubscriptionService (StoreKit -> Google Play Billing).
 * - BillingClient v7+ (queryProductDetailsAsync, queryPurchasesAsync, acknowledgePurchase)
 * - StateFlow pour exposer l'état (équivalent @Published)
 * - Firebase: Auth / Firestore / Analytics
 */
class SubscriptionService private constructor(
    private val appContext: Context
) : PurchasesUpdatedListener {

    companion object {
        @Volatile private var INSTANCE: SubscriptionService? = null

        fun getInstance(context: Context): SubscriptionService {
            return INSTANCE ?: synchronized(this) {
                INSTANCE ?: SubscriptionService(context.applicationContext).also { it.init() ; INSTANCE = it }
            }
        }

        // Identifiants produits (Play Console)
        private const val WEEKLY = "com.lyes.love2love.subscription.weekly"
        private const val MONTHLY = "com.lyes.love2love.subscription.monthly"
    }

    // ---- State (équivalents @Published) ----
    private val _products = MutableStateFlow<List<ProductDetails>>(emptyList())
    val products: StateFlow<List<ProductDetails>> = _products.asStateFlow()

    private val _isSubscribed = MutableStateFlow(false)
    val isSubscribed: StateFlow<Boolean> = _isSubscribed.asStateFlow()

    private val _lastPurchasedProduct = MutableStateFlow<ProductDetails?>(null)
    val lastPurchasedProduct: StateFlow<ProductDetails?> = _lastPurchasedProduct.asStateFlow()

    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()

    private val _errorMessage = MutableStateFlow<String?>(null)
    val errorMessage: StateFlow<String?> = _errorMessage.asStateFlow()

    // ---- Infra ----
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)
    private lateinit var billingClient: BillingClient
    private val connected = AtomicBoolean(false)

    // ---- Initialisation ----
    private fun init() {
        // v7+: enablePendingPurchases(PendingPurchasesParams)
        val pendingParams = PendingPurchasesParams.newBuilder()
            .enableOneTimeProducts() // autorise les pending purchases pour les in-app; ok même si on n’en a pas
            .build()

        billingClient = BillingClient.newBuilder(appContext)
            .setListener(this)
            .enablePendingPurchases(pendingParams)
            .build()

        connectIfNeeded()
        clearPendingErrors()
        // Monitoring périodique comme dans ta version Swift (toutes les heures)
        startSubscriptionStatusMonitoring()
    }

    private fun connectIfNeeded(onReady: (() -> Unit)? = null) {
        if (connected.get()) { onReady?.invoke(); return }

        billingClient.startConnection(object : BillingClientStateListener {
            override fun onBillingSetupFinished(result: BillingResult) {
                if (result.responseCode == BillingClient.BillingResponseCode.OK) {
                    connected.set(true)
                    // Charger produits & achats existants
                    scope.launch {
                        loadProductsInternal()
                        restorePurchasesInternal()
                        // Statut depuis Firestore (cohérence multi-appareils)
                        checkSubscriptionStatusFromFirestore()
                    }
                    onReady?.invoke()
                } else {
                    setError(appContext.getString(
                        R.string.error_billing_setup_failed, result.debugMessage ?: "Unknown"
                    ))
                }
            }

            override fun onBillingServiceDisconnected() {
                connected.set(false)
                // Reconnexion soft
                scope.launch {
                    delay(1500)
                    connectIfNeeded()
                }
            }
        })
    }

    // ---- API publique (équivalent Swift) ----
    fun loadProducts() {
        scope.launch { loadProductsInternal() }
    }

    fun purchase(activity: Activity, product: ProductDetails) {
        scope.launch {
            runCatching {
                val offerToken = chooseOfferToken(product)
                    ?: throw IllegalStateException("Aucune offre disponible pour ${product.productId}")

                val productParams = BillingFlowParams.ProductDetailsParams.newBuilder()
                    .setProductDetails(product)
                    .setOfferToken(offerToken)
                    .build()

                val flowParams = BillingFlowParams.newBuilder()
                    .setProductDetailsParamsList(listOf(productParams))
                    .build()

                withContext(Dispatchers.Main) {
                    _isLoading.value = true
                    val result = billingClient.launchBillingFlow(activity, flowParams)
                    if (result.responseCode != BillingClient.BillingResponseCode.OK) {
                        _isLoading.value = false
                        setError(appContext.getString(
                            R.string.error_launch_purchase, result.debugMessage ?: "Unknown"
                        ))
                    }
                }
            }.onFailure {
                _isLoading.value = false
                setError(it.localizedMessage ?: appContext.getString(R.string.error_generic))
            }
        }
    }

    fun restorePurchases() {
        scope.launch { restorePurchasesInternal() }
    }

    suspend fun handleSubscriptionExpired() {
        val uid = Firebase.auth.currentUser?.uid ?: return
        Log.d("SubscriptionService", "Résiliation pour $uid")
        revokeUserSubscription(uid)
        withContext(Dispatchers.Main) { _isSubscribed.value = false }
        updateSubscriptionStatusInFirestore(false)
    }

    fun startSubscriptionStatusMonitoring() {
        // Vérif toutes les heures (comme Timer.scheduledTimer)
        scope.launch {
            while (isActive) {
                delay(3_600_000L)
                checkSubscriptionStatusFromFirestore()
            }
        }
        Log.d("SubscriptionService", "Monitoring des abonnements démarré")
    }

    // ---- PurchasesUpdatedListener ----
    override fun onPurchasesUpdated(result: BillingResult, purchases: MutableList<Purchase>?) {
        Log.d("SubscriptionService", "onPurchasesUpdated: code=${result.responseCode}")
        if (result.responseCode == BillingClient.BillingResponseCode.OK && purchases != null) {
            scope.launch {
                purchases.forEach { handlePurchase(it) }
            }
        } else {
            _isLoading.value = false
            when (result.responseCode) {
                BillingClient.BillingResponseCode.USER_CANCELED -> {
                    // Annulation utilisateur – pas d’erreur UI nécessaire
                }
                else -> setError(appContext.getString(
                    R.string.error_purchase_failed, result.debugMessage ?: "Unknown"
                ))
            }
        }
    }

    // ---- Interne : chargement produits ----
    private suspend fun loadProductsInternal() {
        _isLoading.value = true
        val query = QueryProductDetailsParams.newBuilder()
            .setProductList(
                listOf(
                    QueryProductDetailsParams.Product.newBuilder()
                        .setProductId(WEEKLY)
                        .setProductType(BillingClient.ProductType.SUBS)
                        .build(),
                    QueryProductDetailsParams.Product.newBuilder()
                        .setProductId(MONTHLY)
                        .setProductType(BillingClient.ProductType.SUBS)
                        .build()
                )
            )
            .build()

        val productsDeferred = CompletableDeferred<List<ProductDetails>>()
        billingClient.queryProductDetailsAsync(query) { br, list ->
            if (br.responseCode == BillingClient.BillingResponseCode.OK) {
                productsDeferred.complete(list)
            } else {
                productsDeferred.completeExceptionally(
                    IllegalStateException("queryProductDetailsAsync: ${br.debugMessage}")
                )
            }
        }

        // Timeout de sécurité (30s) comme ta version Swift
        val details = withTimeoutOrNull(30_000L) { productsDeferred.await() }
        if (details == null) {
            _isLoading.value = false
            setError(appContext.getString(R.string.error_products_timeout))
            return
        }

        _products.value = details
        if (details.isEmpty()) {
            setError(appContext.getString(R.string.error_no_products))
        }
        _isLoading.value = false
    }

    // ---- Interne : restauration / acknowledge ----
    private suspend fun restorePurchasesInternal() {
        if (!connected.get()) connectIfNeeded()
        val params = QueryPurchasesParams.newBuilder()
            .setProductType(BillingClient.ProductType.SUBS)
            .build()

        val result = billingClient.queryPurchasesAsync(params)
        val purchases = result.purchasesList ?: emptyList()
        Log.d("SubscriptionService", "restore: ${purchases.size} purchase(s)")

        purchases.forEach { purchase ->
            when (purchase.purchaseState) {
                Purchase.PurchaseState.PURCHASED -> {
                    handlePurchase(purchase) // fera acknowledge si nécessaire
                }
                Purchase.PurchaseState.PENDING -> {
                    // Informer l’utilisateur qu’un achat est en attente s’il faut
                }
                else -> Unit
            }
        }

        // Met à jour l’UI à partir de Firestore (source de vérité distante)
        checkSubscriptionStatusFromFirestore()
    }

    private suspend fun handlePurchase(purchase: Purchase) {
        val tokenMasked = purchase.purchaseToken.take(8) + "…"
        Log.d("SubscriptionService", "handlePurchase: state=${purchase.purchaseState}, token=$tokenMasked")

        if (purchase.purchaseState != Purchase.PurchaseState.PURCHASED) return

        // 1) Acknowledge obligatoire
        if (!purchase.isAcknowledged) {
            val ackResult = CompletableDeferred<BillingResult>()
            val params = AcknowledgePurchaseParams.newBuilder()
                .setPurchaseToken(purchase.purchaseToken)
                .build()
            billingClient.acknowledgePurchase(params) { br ->
                ackResult.complete(br)
            }
            val br = ackResult.await()
            if (br.responseCode != BillingClient.BillingResponseCode.OK) {
                setError(appContext.getString(R.string.error_acknowledge_failed, br.debugMessage ?: "Unknown"))
                return
            }
        }

        // 2) (Optionnel mais recommandé) Validation serveur du token (Cloud Function)
        // try { validateGooglePurchaseOnServer(purchase) } catch (_: Exception) {}

        // 3) Mise à jour locale & Firestore
        Firebase.analytics.logEvent("abonnement_reussi", null)
        withContext(Dispatchers.Main) {
            _isSubscribed.value = true
            _isLoading.value = false
        }

        // Marquer comme "direct" puis push statut isSubscribed
        markSubscriptionAsDirect()
        updateSubscriptionStatusInFirestore(true)

        // Mémoriser le ProductDetails correspondant pour l’UI
        val pd = products.value.firstOrNull { it.productId in purchase.products }
        _lastPurchasedProduct.value = pd
    }

    // ---- Firestore helpers ----
    private fun updateSubscriptionStatusInFirestore(isSub: Boolean) {
        val uid = Firebase.auth.currentUser?.uid ?: return
        Firebase.firestore.collection("users").document(uid)
            .update(
                mapOf(
                    "isSubscribed" to isSub,
                    "subscriptionExpiredAt" to if (isSub) FieldValue.delete() else com.google.firebase.Timestamp.now()
                )
            )
            .addOnSuccessListener { Log.d("SubscriptionService", "Firestore: isSubscribed=$isSub") }
            .addOnFailureListener { Log.w("SubscriptionService", "Firestore update failed", it) }
    }

    private fun markSubscriptionAsDirect() {
        val uid = Firebase.auth.currentUser?.uid ?: return
        Firebase.firestore.collection("users").document(uid)
            .update(
                mapOf(
                    "subscriptionType" to "direct",
                    "subscriptionPurchasedAt" to com.google.firebase.Timestamp.now()
                )
            )
            .addOnSuccessListener { Log.d("SubscriptionService", "Firestore: subscriptionType=direct") }
            .addOnFailureListener { Log.w("SubscriptionService", "mark direct failed", it) }
    }

    private suspend fun revokeUserSubscription(userId: String) {
        runCatching {
            Firebase.firestore.collection("users").document(userId)
                .update(
                    mapOf(
                        "isSubscribed" to false,
                        "subscriptionExpiredAt" to com.google.firebase.Timestamp.now(),
                        "subscriptionType" to FieldValue.delete()
                    )
                ).awaitUnit()
        }.onSuccess {
            Log.d("SubscriptionService", "Abonnement révoqué pour $userId")
        }.onFailure {
            Log.e("SubscriptionService", "Révocation échouée", it)
        }
    }

    private suspend fun checkSubscriptionStatusFromFirestore() {
        val uid = Firebase.auth.currentUser?.uid ?: return
        runCatching {
            val snap = Firebase.firestore.collection("users").document(uid).get().await()
            val remote = (snap.getBoolean("isSubscribed") ?: false)
            withContext(Dispatchers.Main) { _isSubscribed.value = remote }
        }.onFailure {
            Log.w("SubscriptionService", "check status failed", it)
        }
    }

    // ---- Utilitaires ----
    private fun chooseOfferToken(details: ProductDetails): String? {
        // Simple: prendre la 1ère offer (base plan / offer) disponible
        // Tu peux, si besoin, filtrer via tags/basePlanId pour “weekly”/“monthly”.
        val subs = details.subscriptionOfferDetails ?: return null
        return subs.firstOrNull()?.offerToken
    }

    private fun clearPendingErrors() {
        _errorMessage.value = null
    }

    private fun setError(msg: String) {
        Log.w("SubscriptionService", "UI error: $msg")
        _errorMessage.value = msg
    }
}

// ---- Petites extensions coroutines pour les Tasks Firebase ----
private suspend fun <T> com.google.android.gms.tasks.Task<T>.await(): T =
    suspendCancellableCoroutine { cont ->
        addOnCompleteListener { task ->
            if (task.isSuccessful) cont.resume(task.result, null)
            else cont.resumeWithException(task.exception ?: RuntimeException("Task failed"))
        }
    }

private suspend fun com.google.android.gms.tasks.Task<Void>.awaitUnit() =
    suspendCancellableCoroutine<Unit> { cont ->
        addOnCompleteListener { task ->
            if (task.isSuccessful) cont.resume(Unit, null)
            else cont.resumeWithException(task.exception ?: RuntimeException("Task failed"))
        }
    }
